src/coprocessor/dag/rpn_expr/impl_cast.rs

```diff
@@ -40,7 +43,12 @@ pub fn get_cast_fn(from_field_type: &FieldType, to_field_type: &FieldType) -> Re
         (EvalType::Duration, EvalType::Real) => cast_duration_as_real_fn_meta(),
         (EvalType::Json, EvalType::Real) => cast_json_as_real_fn_meta(),
         _ => return Err(box_err!("Unsupported cast from {} to {}", from, to)),
-    })
+    };
+    // This cast function is inserted by `Coprocessor` automatically,
+    // the `inUnion` flag always false in this situation. Ideally,
+    // the cast function should be inserted by TiDB and pushed down
+    // with all implicit arguments.
+    Ok((func, vec![ScalarArg::Int(0)]))
```

Comment:
This is very inefficient (since there are dynamic allocations and a big type `ScalarArg`). In fact we only use this inUnion flag in very rare cases. Can we make it optional to avoid it in most cases?

Reply:
OK, I will change the `implicit_args` to an `Option` to accomplish this.

Reply:
`Option` doesn't help, does it? When TiDB pushes down functions with implicit parameters like `inUnion`, we have to decode them to `Vec<ScalarValue>`. Of course, temporarily, this `vec![ScalarArg::Int(0)]` is useless. `isUnion` flag isn't used in the cast functions implementation now.

Reply:
Yup. We of course cannot avoid the cost after inUnion is pushed down. However we can avoid the cost when it is not used.

Reply:
Seems we can't use `Option` to avoid this allocation. Thinking about the following scenario:
1. The implicit arguments pushed down are `Some(&[ScalarValue(Some(0))])`.
2. Our `get_cast_fn` returns `None`. 😿